mm-convert-ha-doc (#40)
Azure/DASH: Documentation/High Availability and Scale.md

# High Availability and Scale

## Overview

This document covers the aspect of High Availability and Scalability of

> the Project Sirius and SDN Appliance.

Reviewer: ```suggestion
the project in the SDN Appliance implementation.
```

> The High Availability is achieved via the following concepts:

Reviewer: ```suggestion
High Availability is achieved via the following concepts:
```

-   Active-Passive architecture with Two (2) SDN Appliances
-   Overprovisioning -- same ENI handled by multiple appliance cards
-   Pairing -- each card from one SDN Appliance will have "paired" card
    in second SDN Appliance, with shared/same VIP

Scaling is achieved via the following concepts:

-   Overprovisioning -- same ENI handled by multiple appliance cards,
    each card with different VIP
-   Flow splitting -- TOR (or source node) splitting traffic equally
    across different VIPs

# Requirements

1. 0 downtime planned failover
1. <2 sec downtime unplanned failover
1. Ability to resume connections in event of both unplanned and planned
    failover

> 1. Do not have complicated hashing logic in switches to ensure packets

Reviewer: ```suggestion
1. No complicated hashing logic in switches to ensure packets
```

    land on right appliance
1. If appliance receives a valid packet, it must not drop it due to
    flow replication delays

> 1. Ensure both inbound and outbound packets take same appliance for a

Reviewer: ```suggestion
1. Ensure both inbound and outbound packets transit the same appliance for a
```

    given flow

## Architecture

### High-level architecture of Rack Design for High Availability and Scale

![hl-arch-rack-design-ha](images/has/hl-arch-rack-design-ha.png)

**Considerations**

1. Active/Active hardware design
1. Active/Passive ENI design (software)
1. Two TORs, Cross Wire design
1. Each card connected to both TORs for availability and scale

> 1. Both SDN Appliances are still available in case of single TOR

Reviewer: ```suggestion
1. Both SDN Appliances remain available in case of single TOR
```

1. Paired cards announce same set of VIPs
1. ENI-based (not card based) flow replication (direction of flow replication: "Active ENI" -\> "Passive ENI")
1. Single ENI is programmed on multiple cards (each card with different VIP)
1. TOR (or source node where VM is located) performs traffic load balancing/splitting/sharding for 
selected ENI across VIPs of the cards on which this ENI is provisioned

## High Availability Architecture

### Active-Passive model

Active-Passive model will ensure high availability in case either (1)
one of the TOR/SDN Appliance dies, or (2) single card dies.

**Setup details**

1. Each card from "SDN Appliance 1" will have "paired" card from "SDN Appliance 2"
1. "Paired" cards will be serving same ENI with exact same policies setup on each
1. "Paired" cards will be continuously replicating active flows from Active card to Passive card
1.  Both cards will be announcing same VIP via BGP
1. "SDN Appliance 1" will be in Active mode (announcing preferred, shorter path to itself thru BGP)
1. "SDN Appliance 2" will be in Passive mode (announcing less preferred, longer path to itself thru BGP)

**Normal traffic pattern**

> Normal traffic pattern for ENIs handled by cards will be going always thru "SDN Appliance 1" (Active one).

Reviewer: ```suggestion
Normal traffic pattern for ENIs handled by cards will always transit through "SDN Appliance 1" (Active one).
```

**Failure mode**

In case of failure the BGP routes from "SDN Appliance 1" (previously active) will be withdrawn and TOR will prefer 

> "SDN Appliance 2" and redirect traffic there, ensuring continuous traffic and uninterrupted customer experience.

Reviewer 2: Is the service truly uninterrupted? Isn't there a finite time delay to detect the failure, then withdraw routes? If there is a "short" interruption, we should state some objective, e.g. 90% probability of restoration in under XX time span.

Author: Question for @mzms.

### Pairing

Cards between "SDN Appliance 1" and "SDN Appliance 2" will be paired with each other to created described "Active-Passive" model.

Control plane will be responsible for creating "pairing" relationship between cards -- select which cards create a pair.

### ENI Policy configuration

"Paired" cards will be configured (by control plane) with same ENI and same policy.

Control plane will be responsible for configuring same ENI and same policy across both paired cards. 
To clarify - no replication of ENI policy is required by card -- control plane will do this.

### Flow replication

Once "pairing" relationship is established, flows will need to start being replicated and synced.

> Cards are responsible for replicating and syncing flows across cards in "paired" relationship.

Reviewer 2: How do cards "talk" to each other? 1. What is the physical path? 2. Is there supposed to be a standard protocol for doing this or is it up to the vendors to invent their own?

Reviewer: @mzms - how do the cards 'talk' to each other?

Author: Question for @mzms.

**Important consideration**

In case of outage (ex. entire SDN Appliance not available for longer period of time), the "Pairing" 
relationship might be changed by control plane.

In this case control plane will:

1. Withdraw "pairing" relationship from card (unpair the card)

1. Establish new "pairing" relationship to another card (pair different card)

The original card (which is currently in active state) might and most likely will still continue to receive 
traffic when the pairing relationship will change.

Once new pairing is established, the flow transfer/sync should start. New card will become "passive" from the point of view of 
traffic and to ensure no outage happens - **must not become "active" until all flows are fully synced with original card**.

### Overprovisioning

Same ENI will be handled by multiple cards

## Scalability

Overprovisioning and flow splitting will provide capability of possibly "infinite" CPS as well as "infinite" 
bandwidth for customers, as all the connections will be distributed across multiple cards.

> Control plan will be able to provision different number of cards depending on customer needs for scale.

Reviewer 2: typo - "Control plan" => "Control plane"

Author: fixed

> Control plan will be able to provision different number of cards depending on customer needs for scale.

Reviewer: ```suggestion
Control plane will be able to provision different number of cards depending on customer needs for scale.
```

### Overprovisioning

Single ENI will be provisioned on multiple cards in a single SDN Appliance. Same policy (with exemption of card 
VIP which will different) will be setup on each card.

Each card will be announcing different VIP.

TOR (or source side) will be responsible for splitting (spreading) traffic going thru SDN Appliances across multiple 
VIPs to ensure traffic is equally distributed across all the overprovisioned cards.

In addition, for the purpose of High Availability (as described in previous section), same ENI will be 
also setup on "paired" cards on secondary SDN Appliance.

### Flow splitting

The goal is to ensure that ECMP or any other mechanism will ensure that any set of flows that were active and synced actually 
end up on the passive node. We do not want a ECMP or other mechanism to land a different set of flows that are already synced.

Flow splitting will therefore be done either by "intelligent" TOR or directly on a source based on stable hashing, or directly on the source node (where VM is).

As single ENI will be handled by multiple VIPs (overprovisioned) -- example 23.0.0.1, 23.0.0.2, 23.0.0.3, the TOR will 

> equally rewrite destination address to ensure similar outcome as "ECMP" protocol (with additionalexplicit destination address rewrite).

Reviewer 2: typo "additionalexplicit" => "additional explicit"

Author: fixed

> equally rewrite destination address to ensure similar outcome as "ECMP" protocol (with additionalexplicit destination address rewrite).

Reviewer: ```suggestion
equally rewrite destination address to ensure similar outcome as "ECMP" protocol (with additional explicit destination address rewrite).
```

## Scenarios

### Single TOR Failure

What happens

-   Single TOR becomes inaccessible

How traffic pattern changes

> -   SDN Applinace behind this TOR is still accessible thru second TOR

Reviewer 2: typo "Applinace" => "Appliance"

Author: fixed

Author: fixed

> -   SDN Applinace behind this TOR is still accessible thru second TOR

Author: ```suggestion
-   SDN Appliance behind this TOR is still accessible thru second TOR
```

1. Algorithm finishes. Flow replication has been completed.

#### Notes

-   The above algorithm ensures two sync are happening in parallel:

    -   Real-time immediate sync of new changes (which will remain
        active after "Perfect Sync" is completed)

    -   Sync of existing connections (happening during "Perfect Sync")

-   As the sync is happening, new connections as well as changes in state of existing connections 
(irrespectively of color) is being replicated immediately in real-time (outside of the sync algorithm).

-   It is possible a connection will end (FIN) and will result in the primary card removing flow and immediately 
sending the new connection state change to close the connection to paired devices 
(before the existing connection was even journaled/synced to the paired device). 
To deal with this possibility on the paired device side, if the connection does not already 
exist in its table then this update message should be ignored.

-   When pairing is re-established, it is recommended that the secondary card for simplicity to empty entire state of 
the flow table. This will allow it to receive clean state.

<!-- 
# References

> -   Visio Diagram -

Reviewer: Remove link please

Author: Removed references section

Author: Done

    <https://microsoft.sharepoint-df.com/teams/ProjectSirius/Shared%20Documents/General/Documentation/Design/MSFT%20Topology/AzureDedicated_Sirius_Diagrams.vsdx?web=1>

-   Reduced Tuple support -
    <https://microsoft.sharepoint.com/teams/Aznet/Engineering%20Q%20%20Z/VNET/Sirius/ReducedTupleSupportInVFP.docx?web=1>

> -->

Reviewer: My concern with the drawings is that they have the name 'Sirius' embedded into them. @mzms do we have the .svg of the drawings in order to remove the 'Sirius' reference?

Author: Let me add @mzms to the reviewers.

Author: Question for @mzms.

### Notes

-   The above algorithm ensures two sync are happening in parallel:

    -   Real-time immediate sync of new changes (which will remain
        active after "Perfect Sync" is completed)

    -   Sync of existing connections (happening during "Perfect Sync")

-   As the sync is happening, new connections as well as changes in
    state of existing connections (irrespectively of color) is being
    replicated immediately in real-time (outside of the sync algorithm).

-   It is possible a connection will end (FIN) and will result in the
    primary card removing flow and immediately sending the new
    connection state change to close the connection to paired device
    (before the existing connection was even journaled/synced to the
    paired device). To deal with this possibility on the paired device
    side, if the connection does not already exist in its table then
    this update message should be ignored.

-   When pairing is re-established, it is recommended that the secondary
    card for simplicity to empty entire state of the flow table. This
    will allow it to receive clean state.

# References

-   Visio Diagram -

>     <https://microsoft.sharepoint-df.com/teams/ProjectSirius/Shared%20Documents/General/Documentation/Design/MSFT%20Topology/AzureDedicated_Sirius_Diagrams.vsdx?web=1>

Reviewer 2: These links aren't useful to the community, suggest you get the proper diagrams in the repo and point to them, or else put "TBD" or something.

Author: @chrispsommers  I was still working on the article. Now it should be ready.
